[FIX] Fixing #207 - Skipped. Expected quality not found!

smeggingsmegger opened the pull request:
**In raising this pull request, I confirm the following (please check boxes):**

- [X] I have read and understood the [contributors guide]().
- [X] I have checked that another pull request for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**My familiarity with the project is as follows (check one):**

- [ ] I have never used the project.
- [ ] I have used the project briefly.
- [X] I have used the project extensively, but have not contributed previously.
- [ ] I am an active contributor to the project.

**If the project maintainer has any additional requirements, they will be listed here.**

- No additional requirements.

---

This is a quick and dirty attempt to fix the issue where no videos can be downloaded. The regex being used to parse the HTML for grabbing the video list no longer works. I've modified it to pull in the JSON being passed to the `videojs-setup-data` attribute.

It's a bit more complex than the old method but it is working for me and I hope it is useful for others.

@nishad please feel free to use as much or as little of this fix as you wish.

Thanks for the wonderful and useful utility!

brandon-lile commented on src/udemy_dl/udemy_dl.py:
```diff
@@ -233,8 +234,30 @@ def extract_lecture_url(course_id, lecture_id, asset_type):
+    # The returned HTML has escaped characters in it. This is a really quick
+    # (Probably wrong) way to clean them up.
+    clean_html = json_source['view_html'].replace('&quot;', '"')\
+                                         .replace('\\u0026', '&')\
```
Any chance you thought about using decode or something similar instead of manually searching for the unicode characters?

I.E: `decoded_html = json.source['view_html'].decode('utf-8')`

I would recommend using Beautiful soup for converting the HTML as well, but that is currently not a dependency, and it seems a little overkill to introduce it just for a few cases

smeggingsmegger replied:
Great suggestion!

My earlier code was definitely "quick and dirty". I've trimmed it down to a decode and a .replace of "&quot;" and it's still working.

I, like you, love Beautiful Soup but you are right that it's a big library to include. I definitely didn't want to add any dependencies like that for this small fix. Fantastic library though!

smeggingsmegger replied:
It turns out I don't even need to decode the string. Thanks for the suggestion!

kkarimi commented:
+1

dentvii commented:
Tested @smeggingsmegger forked version, got this error
>  Getting course data: 100.00% [====================] Done
> [CRITICAL-458] Cannot download lecture "How to install Numpy Scipy Matplotlib Pandas IPython Theano and TensorFlow": "'str' object has no attribute 'decode'"

smeggingsmegger commented:
@dentvii I changed my original fix at a suggestion from someone else and it turns out I broke it for Python 3. I have reverted that change and it should be working again.

smeggingsmegger commented:
@kkarimi I removed the .decode and it should be working for Python 3 now.

mdja commented:
How to apply this fix to already install version 0.2.2-alpha ?
I installed using command :
sudo pip install https://github.com/nishad/udemy-dl/zipball/master

sbp-lotus commented:
It is not working with Python 3.5 and Python 3.0 doesn't have scripts folder

smeggingsmegger commented:
@sbp-lotus I am very sorry don't understand. My fix doesn't rely on a "scripts" folder that I'm aware of.

I just tested it on Python 3.5 and it's working for me. Travis is passing on 3.5 as well.

Can you be more specific? Perhaps you can share your error output here.

sevakmaheshr commented:
@nishad , please fix this issue and release an update soon.

waiting for a fixed, working script.

thanks for a great script.

smeggingsmegger commented:
@sevakmaheshr I'm afraid I don't have merge access to this repo. If you want to use my fixed version right now you can pip install from my fork:

`pip install -U git+https://github.com/smeggingsmegger/udemy-dl.git@master`

Cheers!

mahdi87gh commented:
Hi, Can you make the exe version for windows users? I'm having the same problem. I'm getting  [CRITICAL-261] Skipped. Expected quality not found!

sevakmaheshr commented:
@smeggingsmegger, thanks a lot for providing the detailed command. I ran it on OS X El Capitan. It downloaded a whole bunch of stuff, built everything, installed lots of things and (not sure about this) replaced my old (non-working) udemy-dl command with your code.

Things seem to be working now. Thanks a lot for getting udemy-dl working again.

Eshpelin commented:
@smeggingsmegger - Extremely sorry for the noob question - but once I have PIP installed from the location, I am not being able to find the package anywhere. PIP Lists does bring it up, but I can't find the package installation directory - let alone run it. Since I am new to this, would love some help or even pointers to what I should do. 

![problem](https://cloud.githubusercontent.com/assets/23078590/25450104/b217061a-2adf-11e7-8534-3f4b5033ef58.JPG)

black-byte commented:
While trying to run the projecj i got error
`C:\Users\Mateusz\AppData\Local\Programs\Python\Python36-32\python.exe C:/Users/Mateusz/Desktop/udm/udemy-dl-dbe/src/udemy_dl/udemy_dl.py
Traceback (most recent call last):
  File "C:/Users/Mateusz/Desktop/udm/udemy-dl-dbe/src/udemy_dl/udemy_dl.py", line 33, in <module>
    from .download import download, DLException, USER_AGENT
ModuleNotFoundError: No module named '__main__.download'; '__main__' is not a package

Process finished with exit code 1
`

Can anyone help ? I'm using Python 3.6

black-byte commented:
It works!. 

Thank you very much :)

supersambo commented:
works perfectly. Thank you!

ThaYoMin commented:
It probably sounds weird, but I have a little newbie question, how do I run this on Python in windows? 'cause every time I write the lines into the console it says "invalid syntax".
Thanks in advance

black-byte commented:
@ThaYoMin 

1. Go to https://www.python.org/ download python
2. Try install JetBrains PyCharm or other IDE from Python. 
3. Compile it. And run. 
If you are using JetBrains you must go to run settings where you can set all data. 

Thats the most easy way i think for newbie.

ThaYoMin commented:
@black-byte 

Okay, thank you!

lukapaunovic commented:
This patch works perfectly, but why it's not yet merged?
No answer from nishad?

black-byte commented:
@ThaYoMin  No problem :)
@lukapaunovic  Maybe he doesnt support it any more.

smeggingsmegger commented:
@lukapaunovic I'm probably going to just fork the project.

ThaYoMin commented:
@smeggingsmegger That'd be great!

kamran052 commented:
@smeggingsmegger  Hell man I am waiting for the window version!

Please do inform me about your version at kamran3252@gmail.com (Pretty Please!)

soyjonathanespinosa commented:
It's working on Centos Server ;), on windows pc is not working, same error of quality.

whitedoor007 commented:
@smeggingsmegger Hey Bro i tried to install by this command `pip install -U git+https://github.com/smeggingsmegger/udemy-dl.git@master` on Mac OS Sierra but it giving me error everytime.

KAPILs-MacBook-Pro:~ KAPILYADAV$ pip install -U git+https://github.com/smeggingsmegger/udemy-dl.git@master
-bash: /Library/Frameworks/Python.framework/Versions/2.7/bin/pip: No such file or directory
KAPILs-MacBook-Pro:~ KAPILYADAV$ pip install -U git+https://github.com/smeggingsmegger/udemy-dl.git@master
-bash: /Library/Frameworks/Python.framework/Versions/2.7/bin/pip: No such file or directory
KAPILs-MacBook-Pro:~ KAPILYADAV$ pip install -U git+https://github.com/smeggingsmegger/udemy-dl.git@master
-bash: /Library/Frameworks/Python.framework/Versions/2.7/bin/pip: No such file or directory
KAPILs-MacBook-Pro:~ KAPILYADAV$ pip install -U git+https://github.com/smeggingsmegger/udemy-dl.git@master
Collecting git+https://github.com/smeggingsmegger/udemy-dl.git@master
  Cloning https://github.com/smeggingsmegger/udemy-dl.git (to master) to /private/var/folders/4j/xmlp4kj145q9807p85nxgbv80000gn/T/pip-JfBKM0-build
Requirement already up-to-date: requests in /Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from udemy-dl==0.2.3a0)
Requirement already up-to-date: unicode-slugify in /Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from udemy-dl==0.2.3a0)
Requirement already up-to-date: colorlog in /Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from udemy-dl==0.2.3a0)
Requirement already up-to-date: future in /Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from udemy-dl==0.2.3a0)
Requirement already up-to-date: python-pptx in /Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from udemy-dl==0.2.3a0)
Collecting six (from unicode-slugify->udemy-dl==0.2.3a0)
  Using cached six-1.10.0-py2.py3-none-any.whl
Requirement already up-to-date: unidecode in /Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from unicode-slugify->udemy-dl==0.2.3a0)
Requirement already up-to-date: lxml>=3.1.0 in /Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from python-pptx->udemy-dl==0.2.3a0)
Collecting Pillow>=2.6.1 (from python-pptx->udemy-dl==0.2.3a0)
  Using cached Pillow-4.1.1-cp27-cp27m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl
Requirement already up-to-date: XlsxWriter>=0.5.7 in /Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from python-pptx->udemy-dl==0.2.3a0)
Requirement already up-to-date: olefile in /Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from Pillow>=2.6.1->python-pptx->udemy-dl==0.2.3a0)
Installing collected packages: udemy-dl, six, Pillow
  Found existing installation: udemy-dl 0.2.3a0
    Uninstalling udemy-dl-0.2.3a0:
Exception:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pip/commands/install.py", line 342, in run
    prefix=options.prefix_path,
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pip/req/req_set.py", line 778, in install
    requirement.uninstall(auto_confirm=True)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pip/req/req_install.py", line 754, in uninstall
    paths_to_remove.remove(auto_confirm)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pip/req/req_uninstall.py", line 115, in remove
    renames(path, new_path)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pip/utils/__init__.py", line 267, in renames
    shutil.move(old, new)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/shutil.py", line 300, in move
    rmtree(src)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/shutil.py", line 252, in rmtree
    onerror(os.remove, fullname, sys.exc_info())
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/shutil.py", line 250, in rmtree
    os.remove(fullname)
OSError: [Errno 13] Permission denied: '/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/udemy_dl-0.2.3a0-py2.7.egg-info/dependency_links.txt'
KAPILs-MacBook-Pro:~ KAPILYADAV$

iyakexi commented:
This master version doesn't work for me 
pip install -U git+https://github.com/smeggingsmegger/udemy-dl.git@master.

I use Anaconda 4.3.1

dijonyummy commented:
@smeggingsmegger Thanks man! Its working for me now.

AnmolTomer commented:
Works amazingly. Very helpful thanks a ton.

driz commented:
this fixed the problem, please merge @nishad

edoardohorse commented:
pip install -U git+https://github.com/smeggingsmegger/udemy-dl.git@master

works on Windows too

kamran052 commented:
@edoardohorse Link is not opening

dagangwood163 commented:
"pip install -U git+https://github.com/smeggingsmegger/udemy-dl.git@master" works on Windows 10 with Anaconda and Python 3.5 installed.

rogerillu commented:
Thanks a lot! It worked for me too!!

NANO-Chile commented:
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-Bwbwbz/Pillow/

X901 commented:
@smeggingsmegger 
Thank you but for me I had problems 

 I got error message because the pip is not on last version 
and when I write `pip install -U pip`
I go same error

when I use 
`sudo pip install -U pip`

I could update the pip to last version 

but when I write 
`pip install -U git+https://github.com/smeggingsmegger/udemy-dl.git@master
`

I got error file access denial

when I use  
`sudo pip install -U git+https://github.com/smeggingsmegger/udemy-dl.git@master`

it finally installed 

but when i try to use `udemy-dl`

I don't got any error message but it won't download anything 

and I see message 
`etree.so requires version 12.0.0 `

and when I write 

```
brew install libxml2
brew install libxslt
brew link libxml2 --force
brew link libxslt --force
```

Finally everything works fine

lukesUbuntu commented:
thanks this fixed my issue

kabotage commented:
Thanks El Jefe! Fixed the error - "[CRITICAL-238] Skipped. Expected quality not found!"

hegdedarsh commented:
Failed with the below error, can you please help me out.









 gcc -pthread -fno-strict-aliasing -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -DNDEBUG -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -fPIC -Isrc/lxml/includes -I/usr/include/python2.7 -c src/lxml/lxml.etree.c -o build/temp.linux-x86_64-2.7/src/lxml/lxml.etree.o -w
    In file included from src/lxml/lxml.etree.c:515:0:
    src/lxml/includes/etree_defs.h:14:31: fatal error: libxml/xmlversion.h: No such file or directory
     #include "libxml/xmlversion.h"
                                   ^
    compilation terminated.
    Compile failed: command 'gcc' failed with exit status 1
    creating tmp
    cc -I/usr/include/libxml2 -c /tmp/xmlXPathInitgjyygF.c -o tmp/xmlXPathInitgjyygF.o
    /tmp/xmlXPathInitgjyygF.c:1:26: fatal error: libxml/xpath.h: No such file or directory
     #include "libxml/xpath.h"
                              ^
    compilation terminated.
    *********************************************************************************
    Could not find function xmlCheckVersion in library libxml2. Is libxml2 installed?
    *********************************************************************************
    error: command 'gcc' failed with exit status 1

    ----------------------------------------
    Rolling back uninstall of lxml
    Command "/usr/bin/python2.7 -c "import setuptools, tokenize;__file__='/tmp/pip-build-i1hQMT/lxml/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-wDVbMf-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-i1hQMT/lxml

mstuttgart commented:
Oh, its works! Thanks @smeggingsmegger :+1:

nishad commented:
@smeggingsmegger Thanks for doing this. I am accepting this pull. However, please don't bump version via pull requests. Usually project maintainer decides and bumps version. This project follows Semantic Versioning 2.0.0 for its versioning.

Runjesh commented:
@X901 thanks your solution worked for me.

tofanelli commented:
@smeggingsmegger ..... should do any extra step after this installation?? didn't worked for me =\

![image](https://user-images.githubusercontent.com/5512922/27603975-529b4562-5b4e-11e7-9e5a-deaa9328827e.png)
![image](https://user-images.githubusercontent.com/5512922/27603977-5653c076-5b4e-11e7-8313-19087483579b.png)


Cheers =)

vikasnewports commented:
@dagangwood163 an you please guide me I have already installed anaconda + python 3.6 but I am failed to do 
you or anyone else guide me to run python script Windows version not working .

shaanpatel commented:
Been facing this problem since the past 3 days. Was able to download content before. Any insight on why I'm facing this problem?
<img width="618" alt="screen shot 2017-07-06 at 12 35 45 pm" src="https://user-images.githubusercontent.com/14263784/27899449-eb89c640-6247-11e7-82f5-b2d445f7dda7.png">

ludcl commented:
@shaanpatel Same issue here... It was working with @smeggingsmegger fixes but it's not working anymore :(

coffeemood commented:
Any fixes so far guys?

mpolox commented:
Hi,  Hello, Im new on Python, and PIP
I basically downloaded Python,  installed PIP,  and GIT, after that I executed:
"python -m pip install -U git+https://github.com/smeggingsmegger/udemy-dl.git@master"  
and everything went ok,  I got:
 "Successfully installed udemy-dl-0.2.3a0"

But as I said,  I'm new on this,  after this  how do I execute/run the udemy-dl ?

thanks in advance

GavMan1995 commented:
try running this instead, for some reason it works for me when I run this `udemy-dl [course-name] -u username -p password`

oreniko26 commented:
didn't work for me on windows. Attaching log files.

[DEBUG-get_json_source-20170904-113347.txt](https://github.com/nishad/udemy-dl/files/1274123/DEBUG-get_json_source-20170904-113347.txt)
[DEBUG-get_json_source-20170904-113348.txt](https://github.com/nishad/udemy-dl/files/1274120/DEBUG-get_json_source-20170904-113348.txt)
[DEBUG-get_json_source-20170904-113349.txt](https://github.com/nishad/udemy-dl/files/1274121/DEBUG-get_json_source-20170904-113349.txt)
[DEBUG-get_course_data-20170904-113346.txt](https://github.com/nishad/udemy-dl/files/1274122/DEBUG-get_course_data-20170904-113346.txt)

smeggingsmegger commented:
@oreniko26 this has long been merged and working. If there is a new problem you should open a new issue.

oreniko26 commented:
Seems like just installing the Udemy-dl module wasn’t enough. Once I
upgraded from the source directly the problem was fixed. Thanks!

On Mon, 4 Sep 2017 at 12:51 Scott Blevins <notifications@github.com> wrote:

> @oreniko26 <https://github.com/oreniko26> this has long been merged and
> working. If there is a new problem you should open a new issue.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nishad/udemy-dl/pull/209#issuecomment-326919395>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AIYZ-3nv7Ju2zUExI5iQwDznEdFuauERks5se8gUgaJpZM4M-aPF>
> .
>
